File: php/src/http.php
<?php

class HTTP1Request {

    public static function parse(&$stream): HTTP1Request {
        [$method, $url, $version] = parse_first_line($stream);
        $headers = parse_headers($stream);
        $body = parse_body($stream, $headers);
    
        return new HTTP1Request($method, $url, $version, $headers, $body);
    }

    public string $method;
    public string $url;
    public string $version;
    public array $headers;
    public string $body;

    public function __construct($method, $url, $version, $headers, $body) {
        $this->method = $method;
        $this->url = $url;
        $this->version = $version;
        $this->headers = $headers;
        $this->body = $body;
    }

    public function send(&$stream): void {
        if(stream_socket_sendto($stream, "$this->method $this->url $this->version\r\n") === false) throw new Exception("HTTP1Request send stream_socket_sendto failed.");
        foreach($this->headers as &$header) {
            if(stream_socket_sendto($stream, "{$header['key']}: {$header['value']}\r\n") === false) throw new Exception("HTTP1Request send stream_socket_sendto failed.");
        }
        if(stream_socket_sendto($stream, "\r\n") === false) throw new Exception("HTTP1Request send stream_socket_sendto failed.");
        if($this->body && stream_socket_sendto($stream, $this->body) === false) throw new Exception("HTTP1Request send stream_socket_sendto failed.");
    }

    public function print(): void {
        print "\n************************************************ HTTP1Request start\n\n";
        print "$this->method $this->url $this->version\r\n";
        foreach($this->headers as &$header) {
            print $header['key'].": ".$header['value']."\r\n";
        }
        print "\r\n";
        if($this->body) print $this->body."\n";
        print "\n************************************************ HTTP1Request end\n\n";
    }

}

class HTTP1Response {

    public static function parse(&$stream): HTTP1Response {
        [$version, $status_code, $status_text] = parse_first_line($stream);
        $headers = parse_headers($stream);
        $body = parse_body($stream, $headers);

        return new HTTP1Response($version, $status_code, $status_text, $headers, $body);
    }

    public string $version;
    public string $status_code;
    public string $status_text;
    public array $headers;
    public string $body;

    public function __construct($version, $status_code, $status_text, $headers, $body) {
        $this->version = $version;
        $this->status_code = $status_code;
        $this->status_text = $status_text;
        $this->headers = $headers;
        $this->body = $body;
    }

    public function send(&$stream): void {
        if(stream_socket_sendto($stream, "$this->version $this->status_code $this->status_text\r\n") === false) throw new Exception("HTTP1Response send stream_socket_sendto failed.");
        foreach($this->headers as &$header) {
            if(stream_socket_sendto($stream, "{$header['key']}: {$header['value']}\r\n") === false) throw new Exception("HTTP1Response send stream_socket_sendto failed.");
        }
        if(stream_socket_sendto($stream, "\r\n") === false) throw new Exception("HTTP1Response send stream_socket_sendto failed.");
        if($this->body && stream_socket_sendto($stream, $this->body) === false) throw new Exception("HTTP1Response send stream_socket_sendto failed.");
    }

    public function print(): void {
        print "\n************************************************ HTTP1Response start\n\n";
        print "$this->version $this->status_code $this->status_text\r\n";
        foreach($this->headers as &$header) {
            print $header['key'].": ".$header['value']."\r\n";
        }
        print "\r\n";
        if($this->body) print $this->body."\n";
        print "\n************************************************ HTTP1Response end\n\n";
    }

}

function parse_first_line(&$stream): array {
    $first_line = stream_get_line($stream, 0, "\r\n");
    if($first_line === false) throw new Exception("parse_first_line stream_get_line failed.");

    $elements = explode(" ", $first_line);
    if(count($elements) < 3) throw new Exception("parse_first_line invalid first line.");
    $third = implode(" ", array_slice($elements, 2));

    return [$elements[0], $elements[1], $third];
}

function parse_headers(&$stream): array {
    $headers = [];
    while(true) {
        $line = stream_get_line($stream, 0, "\r\n");
        if($line === false) throw new Exception("parse_headers stream_get_line failed.");
        if($line === '') break;
        $key = trim(strstr($line, ":", true));
        $value = trim(substr(strstr($line, ":"), 1));
        array_push($headers, ["key" => $key, "value" => $value]);
    }
    return $headers;
}

function parse_transfer_encoding(&$stream): string {
    $body = "";
    while(true) {
        $chunk_header = stream_get_line($stream, 0, "\r\n");
        if($chunk_header === false) throw new Exception("parse_transfer_encoding chunk_header stream_get_line failed.");

        $body .= $chunk_header."\r\n";

        $hex_str = str_contains($chunk_header, ";") ? strstr($chunk_header, ";", true) : $chunk_header;
        if(!preg_match("/^[a-fA-F0-9]+$/", $hex_str)) throw new Exception("parse_transfer_encoding chunk_header invalid chunk size.");
        $chunk_size = hexdec($hex_str);

        if($chunk_size === 0) {
            while(true) {
                $line = stream_get_line($stream, 0, "\r\n");
                if($line === false) throw new Exception('parse_transfer_encoding trailer or last chunk line(\r\n) stream_get_line failed.');
                $body .= $line."\r\n";
                if($line === '') break;
            }
            break;
        } else {
            $chunk_data = stream_get_contents($stream, $chunk_size);
            if($chunk_data === false) throw new Exception("parse_transfer_encoding chunk_data stream_get_contents failed.");

            $body .= $chunk_data;

            $CRLF = stream_get_line($stream, 0, "\r\n");
            if($CRLF !== '') throw new Exception("parse_transfer_encoding no CRLF after chunk data.");

            $body .= "\r\n";
        }
    }
    return $body;
}

function parse_body(&$stream, array &$headers): string {
    $transfer_encoding = false;
    $content_length = 0;

    foreach($headers as &$header) {
        if(strcasecmp($header["key"], "transfer-encoding") === 0) {
            $transfer_encoding = true;
            break;
        } else if(strcasecmp($header["key"], "content-length") === 0) {
            $content_length = intval($header["value"]);
            if($content_length === 0) throw new Exception("parse_body invalid Content-Length header value");
            break;
        }
    }

    $body = "";
    if($transfer_encoding) {
        $body = parse_transfer_encoding($stream);
    } else if($content_length > 0) {
        $body = stream_get_contents($stream, $content_length);
        if($body === false) throw new Exception("parse_body Content-Length stream_get_contents failed.");
    }

    return $body;
}

File: c++/http.cpp
#include "http.hpp"
#include "util.hpp"
#include "url.hpp"

#include <iostream>
#include <sstream>

#define MAX_LINE_LENGTH 1048575

static std::tuple<std::string, std::string, std::string> parse_first_line(Connection& conn) {
    auto line = conn.readline();
    
    std::stringstream first_line{line};
    
    std::string first;
    std::getline(first_line, first, ' ');
    trim(first);
    
    std::string second;
    std::getline(first_line, second, ' ');
    trim(second);
    
    std::string third;
    std::getline(first_line, third);
    trim(third);

    return std::make_tuple(first, second, third);
}

static std::multimap<std::string, std::string> parse_headers(Connection& conn) {
    std::multimap<std::string, std::string> headers;

    while(true) {
        auto line = conn.readline();
        
        auto header_line = line;
        trim(header_line);
        
        if(header_line.empty()) break;
        
        std::stringstream ss{header_line};
        
        std::string key;
        std::getline(ss, key, ':');
        trim(key);
        transform(key.begin(), key.end(), key.begin(), tolower);
        
        std::string value;
        std::getline(ss, value);
        trim(value);
        
        headers.emplace(key, value);
    }

    return headers;
}

static HTTPBody parse_transfer_encoding(Connection& conn) {
    HTTPBody body;
    while (true) {
        auto line = conn.readline();
        auto chunk_size_line = line;
        rtrim(chunk_size_line);
        size_t chunk_size = std::stoul(chunk_size_line, nullptr, 16);

        std::copy(line.cbegin(), line.cend(), std::back_inserter(body));

        if(chunk_size == 0) {
            while(true) {
                auto line = conn.readline();
                std::copy(line.cbegin(), line.cend(), std::back_inserter(body));
                if(line == "\r\n") break;
            }
            break;
        } else {
            std::vector<uint8_t> chunk(chunk_size);
            conn.read(chunk.data(), 1, chunk_size);
            std::copy(chunk.cbegin(), chunk.cend(), std::back_inserter(body));

            auto line = conn.readline();
            std::copy(line.cbegin(), line.cend(), std::back_inserter(body));
        }
    }
    return body;
}

static HTTPBody parse_body(Connection& conn, std::multimap<std::string, std::string>& headers) {
    HTTPBody body;

    size_t body_size = 0;
    bool transfer_encoding = false;
    std::multimap<std::string, std::string>::iterator itr;
    if((itr = headers.find("content-length")) != headers.end()) {
        body_size = std::stol(itr->second);
    } else if((itr = headers.find("transfer-encoding")) != headers.end()){
        std::stringstream ss{itr->second};
        std::string value;
        while(std::getline(ss, value, ','));
        trim(value);
        if(value != "chunked") {
            std::stringstream ss;
            ss << "not implemented transfer-encoding: " << itr->second;
            throw ss.str();
        }
        transfer_encoding = true;
    }
    
    if(transfer_encoding) {
        body = parse_transfer_encoding(conn);
    } else if(body_size > 0) {
        body.resize(body_size);
        conn.read(body.data(), 1, body_size);
    }

    return body;
}

HTTP1Request::HTTP1Request(std::string method, std::string url, std::string version, std::multimap<std::string, std::string> headers, HTTPBody body): method(method), url(url), version(version), headers(headers), body(body) {}

HTTP1Request::HTTP1Request(Connection& conn) {
    auto [method, url, version] = parse_first_line(conn);
    this->method = method;
    this->url = url;
    this->version = version;

    headers = parse_headers(conn);
    body = parse_body(conn, headers);
}

void HTTP1Request::unproxify() {
    std::stringstream ss;
    URL url = URL(this->url);
    ss << url.pathAndQuery();
    this->url = ss.str();
    for(auto itr = headers.begin(); itr != headers.end();) {
        if(std::equal(itr->first.begin(), itr->first.begin() + 5, "proxy")) {
            itr = headers.erase(itr);
        } else {
            itr++;
        }
    }
}

void HTTP1Request::operator>>(Connection& conn) const {
    conn.write("%s %s %s\r\n", method.c_str(), url.c_str(), version.c_str());
    for(auto header: headers) {
        conn.write("%s: %s\r\n", header.first.c_str(), header.second.c_str());
    }
    conn.write("\r\n");
    if(body.size() > 0) conn.write(body.data(), 1, body.size());
    conn.flush();
}

void operator<<(Connection& conn, const HTTP1Request& request) {
    request >> conn;
}

std::ostream& operator<<(std::ostream& out, const HTTP1Request& request) {
    out << request.method << " " << request.url << " " << request.version << "\r\n";
    for(auto& header: request.headers) {
        out << header.first << ": " << header.second << "\r\n";
    }
    out << "\r\n";

    if(!request.body.empty()) {
        out << std::string(request.body.begin(), request.body.end()) << std::endl;
    }

    return out;
}

HTTP1Response::HTTP1Response(std::string version, std::string status_code, std::string status_text, std::multimap<std::string, std::string> headers, HTTPBody body): version(version), status_code(status_code), status_text(status_text), headers(headers), body(body) {}

HTTP1Response::HTTP1Response(Connection& conn) {
    auto [version, status_code, status_text] = parse_first_line(conn);
    this->version = version;
    this->status_code = status_code;
    this->status_text = status_text;
    
    headers = parse_headers(conn);
    body = parse_body(conn, headers);
}

void HTTP1Response::operator>>(Connection& conn) const {
    conn.write("%s %s %s\r\n", version.c_str(), status_code.c_str(), status_text.c_str());
    for(auto header: headers) {
        conn.write("%s: %s\r\n", header.first.c_str(), header.second.c_str());
    }
    conn.write("\r\n");
    if(body.size() > 0) conn.write(body.data(), 1, body.size());
    conn.flush();
}

void operator<<(Connection& conn, const HTTP1Response& response) {
    response >> conn;
}

std::ostream& operator<<(std::ostream& out, const HTTP1Response& response) {
    out << response.version << " " << response.status_code << " " << response.status_text << "\r\n";
    for(auto& header: response.headers) {
        out << header.first << ": " << header.second << "\r\n";
    }
    out << "\r\n";

    if(!response.body.empty()) {
        out << std::string(response.body.begin(), response.body.end()) << std::endl;
    }

    return out;
}
